h2o-genmodel/src/main/java/hex/genmodel/ModelMojoReader.java

```diff
@@ -263,26 +291,95 @@ else if (line.equals("[columns]")) {
       int n_elements = Integer.parseInt(info[0]);
       String domfile = info[1];
       String[] domain = new String[n_elements];
-      BufferedReader br = _reader.getTextFile("domains/" + domfile);
-      try {
-        String line;
-        int id = 0;  // domain elements counter
+
+
+      try (BufferedReader br = _reader.getTextFile("domains/" + domfile)) {
+        States state = States.NEW_CATEGORICAL;
+        StringBuilder stringBuilder = new StringBuilder();
+        int quoteCount = 0;
+        int domainIdx = 0;
+        int c;
+        parsingLoop:
```

Comment:
I believe labels and controls statements are a really big code-smell and should not be used.

Reply:
Do you have any specific way of re-implementing this in mind ? It could be splitted into more object-oriented way, but then I'm afraid:

- Lots of new frames could potentially be created on the stack (or at least we'll be at the mercy of the VM to do the inlining)
- The code could in a way be more spaghetti. Lots of methods and the state machine would not be in one place.

I supposed the code style can be adjusted a little bit at least - for example `break states;` can be replaced with more clear `continue parsingLoop` and the `parsingLoop` itself might be ranemed to be more descriptive as well.

I have no intention on dismissing your comment, but there are more way to do it. We might even use a pre-built state machine lib, which is unwanted in GenModel. Implementing state machine on our own could be beneficial, e.g. the next CSV parser could utilize it as well. I'm just unsure about the performance. Well tweaked FSA doing only the few steps necessary is a guarantee of speed.

Please provide some examples on what would you ideally want to do. I'll also try to think about it.

Reply:
I'm going to discard this whole block in favor of character escaping. So we'll get rid of the handcrafted state machine.